src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -524,11 +545,33 @@ private ValueTask<FlushResult> WriteDataAndTrailersAsync(Http2Stream stream, in
 
             try
             {
-                _headersEnumerator.Initialize(headers);
+                // In the case of the trailers, there is no status header to be written, so even the first call to BeginEncodeHeaders can return BufferTooSmall.
                 _outgoingFrame.PrepareHeaders(Http2HeadersFrameFlags.END_STREAM, streamId);
-                var buffer = _headerEncodingBuffer.AsSpan();
-                var done = HPackHeaderWriter.BeginEncodeHeaders(_hpackEncoder, _headersEnumerator, buffer, out var payloadLength);
-                FinishWritingHeaders(streamId, payloadLength, done);
+                var bufferSize = _headerEncodingBuffer.Capacity;
+                HPackHeaderWriter.HeaderWriteResult done;
+                do
+                {
+                    _headersEnumerator.Initialize(headers);
+                    _headerEncodingBuffer.ResetWrittenCount();
+                    var buffer = _headerEncodingBuffer.GetSpan(bufferSize)[0..bufferSize]; // GetSpan might return more data that can result in a less deterministic behavior on the way headers are split into frames.
+                    done = HPackHeaderWriter.BeginEncodeHeaders(_hpackEncoder, _headersEnumerator, buffer, ref _currentResponseHeadersTotalSize, _maxResponseHeadersTotalSize, out var payloadLength);
+                    if (done == HPackHeaderWriter.HeaderWriteResult.Done)
+                    {
+                        _headerEncodingBuffer.Advance(payloadLength);
+                        SplitHeaderFramesToOutput(streamId, done, isFramePrepared: true);
+                    }
+                    else if (done == HPackHeaderWriter.HeaderWriteResult.MoreHeaders)
+                    {
+                        // More headers sent in CONTINUATION frames.
+                        _headerEncodingBuffer.Advance(payloadLength);
+                        SplitHeaderFramesToOutput(streamId, done, isFramePrepared: true);
+                        FinishWritingHeaders(streamId);
+                    }
+                    else
+                    {
+                        bufferSize *= 2;
```

Comment:
Ignorant question: does increasing the buffer size beyond the capacity cause it to grow?  Will the next `GetSpan` call just notice that the capacity has been exceeded and react appropriately?

Reply:
Yes, but I can double check too.

Reply:
Yes, the tests would fail otherwise